geo/aws.go

```diff
@@ -0,0 +1,121 @@
+package geo
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+)
+
+const (
+	AWSGeoUrl  = "https://ip-ranges.amazonaws.com/ip-ranges.json"
+	AWSGeoFile = "ip-ranges.json"
+)
+
+type AWSProvider struct {
+}
+
+type AWSGeoInfo struct {
+	SyncToken  string      `json:"syncToken"`
+	CreateData string      `json:"createDate"`
+	Prefixes   []AWSIPInfo `json:"prefixes"`
+}
+
+type AWSIPInfo struct {
+	IPPrefix           string `json:"ip_prefix"`
+	Region             string `json:"region"`
+	Service            string `json:"service"`
+	NetworkBorderGroup string `json:"network_border_group"`
+}
+
+var AWSRegionMapping = map[string]string{
+	"us-east-1":      "US",
+	"us-east-2":      "US",
+	"us-west-1":      "US",
+	"us-west-2":      "US",
+	"af-south-1":     "ZA",
+	"ap-east-1":      "HK",
+	"ap-south-1":     "IN",
+	"ap-northeast-1": "JP",
+	"ap-northeast-2": "KR",
+	"ap-southeast-1": "SG",
+	"ap-southeast-2": "AU",
+	"ca-central-1":   "CA",
+	"eu-central-1":   "DE",
+	"eu-west-1":      "IE",
+	"eu-west-2":      "EN",
+	"eu-west-3":      "FR",
+	"eu-south-1":     "IT",
+	"eu-north-1":     "SE",
+	"me-south-1":     "BH",
+	"sa-east-1":      "BR",
+}
+
+func (p AWSProvider) GetLocation(ip string) (*Location, error) {
+	loc, err := getLocationFromAWS(ip)
+	if err != nil {
+		return &emptyLocation, err
+	}
+	return loc, nil
+}
+
+func getLocationFromAWS(ip string) (*Location, error) {
+	if _, err := os.Stat(AWSGeoFile); os.IsNotExist(err) {
+		AWSFile, err := os.Create(AWSGeoFile)
+		if err != nil {
+			return nil, err
+		}
+		defer AWSFile.Close()
+
+		resp, err := http.Get(AWSGeoUrl)
+		if err != nil {
+			return nil, err
+		}
+		defer resp.Body.Close()
+
+		_, err = io.Copy(AWSFile, resp.Body)
```

Comment:
It's better to fetch it to memory, make sure data is valid, and then write to disk. Otherwise you might write partial or invalid data to disk